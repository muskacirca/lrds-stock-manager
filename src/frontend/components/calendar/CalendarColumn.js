import React from 'react'
import moment from 'moment'
import slice from 'lodash/slice'
import concat from 'lodash/concat'
import classnames from 'classnames'

import {
    computeFreeTime
} from './utils/CalendarUtils'

import {
    Link
} from 'react-router'

class CalendarColumn extends React.Component {

    constructor(props) {
        super(props)
    }

    renderEventsList(events) {

        let slicedArray = slice(events.map((event, key) => {
            return  <li key={"calendar-event-list-" + this.props.dayNumber + "-" + key}>
                        <Link to={"/event/" + event.node.id}>{event.node.name}</Link>
                    </li>
        }), 0, 3);

        return events.length > 3 ? concat(slicedArray, "...") : slicedArray
    }

    renderOccupation(events) {

        let freeTime = "";
        if(events) {
            freeTime = computeFreeTime(moment(), events);
        }
        return [<div key="freetime-">{freeTime}</div>]
    }

    handleEventBadgeClick() {

    }

    renderMoreButton() {
        return  <div className="mobile-hide calendar-event-plus-wrapper">
                    <div className="pointer calendar-event-plus"
                         onClick={this.handleEventBadgeClick.bind(this)}>
                        <i className="fa fa-plus-circle" aria-hidden="true"/>
                    </div>
                </div>
    }

    render() {

        let dayNumber = this.props.dayNumber;
        let now = moment();
        
        let events = this.renderEventsList(this.props.events);
        
        let labelClass = classnames({"calendar-days-tr-up": true}, 
                                    {"red": dayNumber == now.date()
                                                && this.props.defaultDate.month() == now.month()
                                                && this.props.defaultDate.year() == now.year()});


        let moreButton = this.props.events.length > 3 ? this.renderMoreButton() : null;

        return  <div className="calendar-row-content calendar-days">
                    <div className={labelClass}>
                        {this.props.label ? this.props.label + ' ' + dayNumber : dayNumber}
                    </div>
                    <div className="calendar-days-content">

                        <ul className="mobile-hide">{events}</ul>
                        {moreButton}
                        <div className="calendar-event-count-wrapper hidden-md hidden-sm hidden-lg">
                            <div className="pointer calendar-event-count"
                                onClick={this.handleEventBadgeClick.bind(this)}>
                                <span className="badge">{events.length > 0 ? events.length : null}</span>
                            </div>
                        </div>
                    </div>
                </div>
    }


}

CalendarColumn.propTypes = {
    label: React.PropTypes.string,
    displayType:  React.PropTypes.oneOf(['month', 'week'])
};

CalendarColumn.defaultProps = {
    displayType: "month"
};

export default CalendarColumn
