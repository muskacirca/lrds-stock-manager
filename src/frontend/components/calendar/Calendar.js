import React from 'react'
import moment from 'moment';

import {
    getFirstDayOfMonth
} from './utils/DateUtils'

import classnames from 'classnames'

const DAYS_NAME = ['Mon','Tue','Wed','Thu','Fri','Sat','Sun'];

import CalendarColumn from './CalendarColumn'

class Calendar extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            defaultDate: moment(this.props.defaultDate)
        }
    }

    getDaysOfWeek() {
        return this.state.defaultDate.format("ddd");
    }

    getFirstDayOfWeek() {

        let day = this.state.defaultDate.format("e");
        return  this.state.defaultDate.subtract(day, 'days');
    }


    getDayLabel(dayOfWeek) {

        let day = this.state.defaultDate.format("e");
        let firstDayOfWeek = this.state.defaultDate.subtract(day, 'days');
        return firstDayOfWeek.add(dayOfWeek, 'days')
    }

    getDaysInMonth() {
        return this.state.defaultDate.daysInMonth();
    }

    renderCalendar(blankDays) {

       return this.props.displayType === "month"
        ? this.renderCalendarMonth(blankDays)
           : this.renderCalendarWeek()
    }

    renderCalendarWeek() {

        let firstDayOfWeek = this.getFirstDayOfWeek().format("D");
        let dayLabel = this.getFirstDayOfWeek().format("dddd");

        let weekRow = [];
        for(let day=0; day<7; day++) {

            let eventsOfTheDay = this.findEventsByDay(this.props.events, firstDayOfWeek);
            console.log("firstDayOfWeek : " + JSON.stringify(firstDayOfWeek));
            let dayLabel = this.getDayLabel(day).format("dddd");

            weekRow.push(<CalendarColumn defaultDate={this.state.defaultDate}
                                         label={dayLabel}
                                         dayNumber={firstDayOfWeek}
                                         events={eventsOfTheDay}
                                         key={"calendar-days-" + firstDayOfWeek} />)


            firstDayOfWeek++
        }

        return <div className="calendar-week-row">{weekRow}</div>
    }

    renderCalendarMonth(blankDays) {
        let dayNumber = 1;
        let calendar = [];

        for(let c=0; c<6; c++) {

            if(dayNumber > this.getDaysInMonth()) break;

            let weekRow =  this.renderWeekRow(dayNumber, blankDays)
            calendar.push(<div className="calendar-month-row" key={"calendar-row-" + c}>{weekRow.html}</div>)

            blankDays = [];
            dayNumber = weekRow.newDayNumber
        }

        return calendar
    }

    renderWeekRow(initialDayNumber, blankDays) {

        let dayNumber = initialDayNumber;
        let counter = blankDays.length === 0 ? 0 : blankDays.length;

        for(var c=counter ; c < 7; c++) {
            if(dayNumber > this.getDaysInMonth()) break;

            let eventsOfTheDay = this.findEventsByDay(this.props.events, dayNumber);

            blankDays.push(<CalendarColumn defaultDate={this.state.defaultDate}
                                           dayNumber={dayNumber}
                                           events={eventsOfTheDay}
                                           key={"calendar-days-" + dayNumber} />);
            dayNumber++
        }



        return {html: this.fillWithBlankDays(blankDays), newDayNumber: dayNumber}
    }


    fillWithBlankDays(aRowOfDays) {

        while(aRowOfDays.length != 7) {
            aRowOfDays.push(<div className="calendar-row-content calendar-blank-days" key={"calendar-blank-days-" + aRowOfDays.length}></div>)
        }
        return aRowOfDays
    }

    renderCalendarHeader() {
        if(this.props.displayType == "month") {
            let days = DAYS_NAME.map(day => {
                return  <div className="calendar-row-content calendar-days-header" key={"calendar-days-header-" + day}>
                            {day}
                        </div>
            });

            return  <div className="calendar-row calendar-month-header">{days}</div>
        }
    }

    renderBlankDays() {
        let blankDays = [];
        let i = 0;
        while (getFirstDayOfMonth(this.state.defaultDate, "ddd") != DAYS_NAME[i]) {
            blankDays[i] = <div className="calendar-row-content calendar-blank-days" key={"calendar-blank-days-" + i}></div>
            i++;
        }

        return blankDays
    }

    findEventsByDay(events, dayNumber) {

        return events != undefined
            ?   events.filter(event => {
                    return moment(event.node.startDate).dates() == dayNumber
                })
            : []
    }

    componentWillReceiveProps(newprops) {
        this.setState({defaultDate: newprops.defaultDate})
    }

    render() {

        let calendarHeaderDays = this.renderCalendarHeader();
        let blankDays = this.renderBlankDays();
        let calendar = this.renderCalendar(blankDays);
        
        console.log("this.props.displayType : " + JSON.stringify(this.props.displayType));
        
        let calendarClass = classnames({"calendar-month-view": this.props.displayType == "month"}, 
                                        {"calendar-week-view": this.props.displayType == "week"});

        return  <div className={calendarClass}>
                    {calendarHeaderDays}
                    {calendar}
                </div>

    }


}

Calendar.propTypes = {
    events: React.PropTypes.array,
    displayType:  React.PropTypes.oneOf(['month', 'week'])
};

Calendar.defaultProps = {
    defaultDate: moment(),
    displayType: "month"
};

export default Calendar
