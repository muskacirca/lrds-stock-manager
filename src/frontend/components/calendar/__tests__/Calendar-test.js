jest.unmock('../Calendar')

import React from 'react';
import Calendar from '../Calendar'
import CalendarColumn from '../CalendarColumn';

import renderer from 'react-test-renderer'

describe('Calendar', () => {

    const defaultDate = new Date(2016, 0, 1, 10, 30, 2)


    it('Display the default date in week view', () => {

        const tree = renderer.create(
            <Calendar defaultDate={defaultDate} displayType="week" />
        ).toJSON();
        
        expect(tree).toMatchSnapshot();
        
    });
    
    
    it('Display the default date in month view ', () => {

        const tree = renderer.create(
            <Calendar defaultDate={defaultDate} displayType="month" />
        ).toJSON();

        expect(tree).toMatchSnapshot();
        
        // let instance = TestUtils.renderIntoDocument(<Calendar defaultDate={defaultDate} />);

        // var calendarHeaderColumns = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'calendar-days-header');
        // expect(calendarHeaderColumns.length).toEqual(7);
        // expect(calendarHeaderColumns[0].textContent).toEqual("Mon");
        // expect(calendarHeaderColumns[1].textContent).toEqual("Tue");
        //
        // var calendarBlankDays = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'calendar-blank-days');
        // expect(calendarBlankDays.length).toEqual(11);
        //
        // expect(CalendarColumn.mock.calls.length).toBe(31);
        //
        // var calendarContent = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'calendar-month-row');
        // expect(calendarContent.length).toEqual(6)
        // expect(calendarContent[0].props.children.length).toEqual(7)
        // expect(calendarContent[1].props.children.length).toEqual(7)
        // expect(calendarContent[2].props.children.length).toEqual(7)
        // expect(calendarContent[3].props.children.length).toEqual(7)
        // expect(calendarContent[4].props.children.length).toEqual(7)
    })
});
