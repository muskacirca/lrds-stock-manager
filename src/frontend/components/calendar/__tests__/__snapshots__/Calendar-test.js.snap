exports[`Calendar Display the default date in month view  1`] = `
<div
  className={undefined}>
  <div
    className="calendar-row calendar-month-header">
    <div
      className="calendar-row-content calendar-days-header">
      Mon
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Tue
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Wed
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Thu
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Fri
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Sat
    </div>
    <div
      className="calendar-row-content calendar-days-header">
      Sun
    </div>
  </div>
  <div
    className="calendar-month-row">
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
  </div>
  <div
    className="calendar-month-row" />
  <div
    className="calendar-month-row" />
  <div
    className="calendar-month-row" />
  <div
    className="calendar-month-row" />
  <div
    className="calendar-month-row">
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
    <div
      className="calendar-row-content calendar-blank-days" />
  </div>
</div>
`;

exports[`Calendar Display the default date in week view 1`] = `
<div
  className={undefined}>
  <div
    className="calendar-week-row" />
</div>
`;
